src/rt/rwlock.rs

```diff
@@ -0,0 +1,243 @@
+use crate::rt::{
+    execution,
+    object::{self, Object},
+    thread, Access, Synchronize,
+};
+
+use std::collections::HashSet;
+use std::sync::atomic::Ordering::{Acquire, Release};
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub(crate) struct RwLock {
+    obj: Object,
+}
+
+#[derive(Debug, PartialEq)]
+enum Locked {
+    Read(HashSet<thread::Id>),
+    Write(thread::Id),
+}
+
+#[derive(Debug)]
+pub(super) struct State {
+    /// A single `thread::Id` when Write locked.
+    /// A set of `thread::Id` when Read locked.
+    lock: Option<Locked>,
```

Comment:
Why does this have to be `Option<Locked>`? Could it be `Locked` and initialized at first use?

Reply:
If no threads are reading or writing the RwLock then it doesn't make sense for it to be in either `Locked::Read` or `Locked::Write`, so I think `None` makes the most sense.

Reply:
Ah, so similar to the following
```rust
#[derive(Debug, PartialEq)]
enum Locked {
    Read(HashSet<thread::Id>),
    Write(thread::Id),
    None,
}
```
I guess `Option<Locked>` makes more sense now but I wonder if the size is any different from `Option<Locked>` compared to `Locked`.

Reply:
Hmm. I don't know about that. I was more focused on the expressiveness of the implementation and not the size impact. I can change it to your suggestion if we're concerned about it. Your suggestion seems pretty expressive too -- "the RwLock can be in one of three states..." makes sense too.

Reply:
No, I am not sure whether it affects the size, just thought about it, we need to verify it with probably something like `std::mem::size_of<Type>()`.